theano/tensor/nnet/tests/test_spatialtf.py

```diff
@@ -0,0 +1,222 @@
+        output = T.sum([wa * Ia, wb * Ib, wc * Ic, wd * Id], axis=0)
+        return output
+
+    def _linspace(start, stop, num):
+        # Theano linspace. Behaves similar to np.linspace
+        start = T.cast(start, theano.config.floatX)
+        stop = T.cast(stop, theano.config.floatX)
+        num = T.cast(num, theano.config.floatX)
+        step = (stop - start) / (num - 1)
+        return T.arange(num, dtype=theano.config.floatX) * step + start
+
+    def _meshgrid(height, width):
+        # This function is the grid generator from eq. (1) in reference [1].
+        # It is equivalent to the following numpy code:
+        #  x_t, y_t = np.meshgrid(np.linspace(-1, 1, width),
+        #                         np.linspace(-1, 1, height))
+        #  ones = np.ones(np.prod(x_t.shape))
+        #  grid = np.vstack([x_t.flatten(), y_t.flatten(), ones])
+        # It is implemented in Theano instead to support symbolic grid sizes.
+        # Note: If the image size is known at layer construction time, we could
+        # compute the meshgrid offline in numpy instead of doing it dynamically
+        # in Theano. However, it hardly affected performance when we tried.
+        x_t = T.dot(T.ones((height, 1)),
+                    _linspace(-1.0, 1.0, width).dimshuffle('x', 0))
+        y_t = T.dot(_linspace(-1.0, 1.0, height).dimshuffle(0, 'x'),
+                    T.ones((1, width)))
+
+        x_t_flat = x_t.reshape((1, -1))
+        y_t_flat = y_t.reshape((1, -1))
+        ones = T.ones_like(x_t_flat)
+        grid = T.concatenate([x_t_flat, y_t_flat, ones], axis=0)
+        return grid
+
+    num_batch, num_channels, height, width = inp.shape
+    theta = T.reshape(theta, (-1, 2, 3))
+
+    # grid of (x_t, y_t, 1), eq (1) in ref [1]
+    out_height = T.cast(T.ceil(height * scale_height), 'int64')
+    out_width = T.cast(T.ceil(width * scale_width), 'int64')
+    grid = _meshgrid(out_height, out_width)
+    # transform a x (x_t, y_t, 1)^t -> (x_s, y_s)
+    t_g = T.dot(theta, grid)
+    x_s = t_g[:, 0]
+    y_s = t_g[:, 1]
+    x_s_flat = x_s.flatten()
+    y_s_flat = y_s.flatten()
+
+    # dimshuffle input to  (bs, height, width, channels)
+    input_dim = inp.dimshuffle(0, 2, 3, 1)
+    input_transformed = _interpolate(
+        input_dim, x_s_flat, y_s_flat,
+        out_height, out_width, border_mode)
+
+    output = T.reshape(
+        input_transformed, (num_batch, out_height, out_width, num_channels))
+    output = output.dimshuffle(0, 3, 1, 2)  # dimshuffle to conv format
+    return output
+
+
+class TestTransformer(unittest.TestCase):
+    def setUp(self):
+        utt.seed_rng()
+
+    def getInputs(self):
+        inp_shape = (np.random.randint(1, 11), 3, np.random.randint(10, 101),
+                     np.random.randint(10, 101))
+        num_images, num_channels, height, width = inp_shape
+
+        inp = np.random.random(inp_shape).astype(config.floatX)
+
+        transform = [[-1, 0, 0],
+                     [0, -1, 0]]
```

Comment:
When I change transform value, I got severe WrongValue errors.

Extend your tests to test with various values of transform (including floating values).

Reply:
I'll add more tests, and use random transformation matrices, to see where the problems may be arising.

Reply:
Hello @joaovictortr don't forget this !